File: backend/app/services/normalize.py
import re

WHITESPACE_RE = re.compile(r"\s+")
FILL_BLANK_RE = re.compile(r"[_×X]{3,}")


def normalize(text: str, *, ignore_whitespace: bool = True) -> str:
    if not text:
        return ""

    result = text.strip()
    if ignore_whitespace:
        result = WHITESPACE_RE.sub("", result)

    result = result.replace("，", ",").replace("。", ".").replace("；", ";")
    result = result.replace("（", "(").replace("）", ")").replace("：", ":")
    return result


def is_fill_blank(text: str) -> bool:
    return bool(FILL_BLANK_RE.search(text))

File: backend-node/src/lib/compare/normalize.ts
const WHITESPACE = /\s+/g

/** 规范化文本：可选去空白，并将中文标点统一为半角形式 */
export function normalize(text: string, ignoreWhitespace: boolean): string {
  if (!text) return ''

  let result = text.trim()
  if (ignoreWhitespace) {
    result = result.replace(WHITESPACE, '')
  }

  return result
    .replace(/，/g, ',')
    .replace(/。/g, '.')
    .replace(/；/g, ';')
    .replace(/（/g, '(')
    .replace(/）/g, ')')
    .replace(/：/g, ':')
}
